oeps/oep-0018-bp-python-dependencies.rst

```diff
@@ -0,0 +1,344 @@
+source Python projects) for ease of understanding and maintenance.
+
+Specification
+=============
+
+The key to successful Python dependency management is to break it down into
+four parts:
+
+1. Identify the different contexts in which dependencies will need to be
+   installed.
+2. For each of these contexts, declare the direct dependencies that will be
+   needed.  Use the least restrictive constraints which should allow pip to
+   install a working set of dependencies.
+3. Auto-generate from the high-level dependencies declaration a complete
+   set of exact package versions that are known to work for a Python
+   `virtualenv`_ created for that context.
+4. Automate updates of the detailed dependencies listing for each context.
+
+.. _virtualenv: https://virtualenv.pypa.io/
+
+Identify Usage Contexts
+-----------------------
+
+The dependencies of Python software are typically installed or run in a
+variety of different contexts over the course of developing and using it.
+The set of dependencies needed to perform a task can easily vary between these
+contexts.  For example:
+
+* Just the standard set of core dependencies for execution on a production
+  server to perform its primary purpose (``base.in``)
+* Additional dependencies which are only needed when optional extra features
+  of a package are desired
+* Assorted testing libraries to run automated test suites (``test.in``)
+* Static code analysis tools to perform code quality checks (``quality.in``)
+* The utilities called directly by a CI server to create and use one or more
+  virtualenvs and report code coverage statistics to a 3rd-party service
+  (``jenkins.in``, ``travis.in``)
+* `Sphinx`_ and other utilities used to generate developer documentation
+  (``doc.in``)
+* Additional utilities needed to perform common development tasks (``dev.in``)
+* Utilities that a particular developer likes to use with a repository, but
+  aren't strictly needed for any of the regular contexts (``private.in``).
+
+.. _Sphinx: http://www.sphinx-doc.org/
+
+Declare Direct Dependencies
+---------------------------
+
+As indicated above, some of the usage contexts have a standard filename used in
+the ``requirements`` directory of an Open edX repository to list dependencies.
+Others will have an appropriate filename custom to that repository's unique
+context.  Each of these is a ``pip``-compatible `requirements file`_ listing
+the direct dependencies needed for that context.  Beyond complying with the
+file format, there are a few guidelines each of these files should follow:
+
+* The file should start with a brief comment explaining the context in which
+  these dependencies are needed.  Examples can be found in the
+  `cookiecutter-django-app`_ repository.
+* Each listed dependency should have a brief end-of-line comment explaining
+  its primary purpose(s) in this context.  These comments typically start at
+  the 27th character, but this is just a convention for consistency with files
```

Comment:
27th!? We live in a 4-space-indent world, and pip-compile goes to the 27th character? :)

Reply:
About the purpose: wouldn't this often be obvious or trivial?  It sounds kind of like requirement comments on every line of a program in an intro programming course: they don't end up being useful.

Reply:
Most of those leading 26 characters are often the package name and optional version constraint.  pip-tools lines up the start of the comments for ease of reading, I figured it just makes sense to match that for consistency.

Unlike imports in a Python file where you can see later in the file what they're used for, seeing something like "singledispatch" in a requirements file tells you very little about why it's there.  The main reason for adding the comments is to make it easier to identify requirements that are no longer needed, which is tough if you're not even sure what half of them do.  I'm thinking things like "celery task broker" for redis and "Used in shoppingcart for extracting/parsing pdf text" for pdfminer; the reason for including as a dependency, not the overview from the package metadata.  I've typically found it very tough to prune out obsolete dependencies without having this kind of comments in the file, but I'm open to arguments against it.

Reply:
It seems like something that would easily get out of date.  What would the comment for "singledispatch" be? "# So that I can do single dispatch"?  If I add it, and put "# used in the foobar manager" in the requirements file, how will that comment get updated if it's also used elsewhere, or if it's no longer used in its original spot?  Wouldn't grepping the source for imports be a better way to understand whether and how packages are being used?

Reply:
About the indentation: looking at the .in files in the edx-platform pull request, on a few lines, 26 isn't enough, and the alignment is bumped out.  I'd choose a larger number that is a multiple of four, and let pip-tools do what it wants. In some ways, it's better to have a clear distinction between the tool-generated and human-generated comments anyway.

Reply:
About the comments themselves:
```
lxml==3.8.0               # XML parser
```
I don't find this useful.  I don't know how to get the useful comments without requiring people to add useless ones like this.

Reply:
I struggled a bit with the `edx-sandbox/shared.in` file comments because it seems like a grab bag of "make this stuff available to instructors in case they want to use it".  So yeah, those ended up being pretty generic, but I think that's more the exception than the rule.

Reply:
I guess I am skeptical that these comments will be accurate and specific enough to be useful in the long run.  But we can give it a try. At least requiring a comment will slow people down from adding too many dependencies! :)